src/librustc/infer/opaque_types/mod.rs

```diff
@@ -326,14 +329,27 @@ impl<'a, 'gcx, 'tcx> InferCtxt<'a, 'gcx, 'tcx> {
                         // There are two regions (`lr` and
                         // `subst_arg`) which are not relatable. We can't
                         // find a best choice.
-                        self.tcx
+                        let context_name = match opaque_defn.origin {
+                            hir::ExistTyOrigin::ExistentialType => "existential type",
+                            hir::ExistTyOrigin::ReturnImplTrait => "impl Trait",
+                            hir::ExistTyOrigin::AsyncFn => "async fn",
+                        };
+                        let msg = format!("ambiguous lifetime bound in `{}`", context_name);
+                        let mut err = self.tcx
                             .sess
-                            .struct_span_err(span, "ambiguous lifetime bound in `impl Trait`")
-                            .span_label(
-                                span,
-                                format!("neither `{}` nor `{}` outlives the other", lr, subst_arg),
-                            )
-                            .emit();
+                            .struct_span_err(span, &msg);
+                        err.span_label(
+                            span,
+                            format!("neither `{}` nor `{}` outlives the other", lr, subst_arg),
+                        );
+                        if let hir::ExistTyOrigin::AsyncFn = opaque_defn.origin {
+                            err.note(
```

Comment:
For consistency with existing diagnostics, it would be better to split this note into each of its parts.
```rust
err.note("multiple unrelated lifetimes are not allowed in `async fn`");
err.note("if you're using argument-position elided lifetimes, consider \
                                switching to a single named lifetime");
```